### PR TITLE
Bug 1876823 - Add action to fetch list of sites that should not be tr…

### PR DESCRIPTION
**Pull request by calumozilla**
…anslated

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1876823

### CODE REVIEW

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt [hunk 1]
```diff
@@ -111,6 +116,44 @@ class TranslationsMiddleware(
...
+     * store via [TranslationsAction.SetNeverTranslateSitesAction] or else dispatches the failure
+     * [TranslationsAction.TranslateExceptionAction].
+     *
+     * @param context Context to use to dispatch to the store.
+     * @param tabId Tab ID associated with the request.
+     */
+    private suspend fun getNeverTranslateSites(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        tabId: String,
+    ) = withContext(Dispatchers.IO) {
```
Comment (ohall-m) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:130]: We recently explored more on this `Dispatchers.IO` over in https://github.com/mozilla-mobile/firefox-android/pull/5409 and a change landed today. Looks like the best recommended path forward is no `withContext` and dispatch once with `scope.launch`.

```suggestion
private suspend fun getNeverTranslateSites(
        context: MiddlewareContext<BrowserState, BrowserAction>,
        tabId: String,
        ) {
        engine.getNeverTranslateSiteList(
```

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 2]
```diff
@@ -104,6 +104,17 @@ internal object TranslationsStateReducer {
                         )
                     }
                 }
+
+                TranslationOperation.FETCH_NEVER_TRANSLATE_SITES -> {
+                    // Reset the error state, and then generally expect
+                    // [TranslationsAction.SetPageSettingsAction] to update state in the
```
Comment (ohall-m) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt:110]: nit:
```suggestion
                    // [TranslationsAction.SetNeverTranslateSitesAction] to update state in the
```

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt [hunk 3]
```diff
@@ -203,6 +203,24 @@ class TranslationsActionTest {
         assertEquals(supportedLanguages, tabState().translationsState.supportedLanguages)
     }
 
+    @Test
+    fun `WHEN a SetNeverTranslateSitesAction is dispatched AND successful THEN update neverTranslateSites`() {
```
Comment (ohall-m) [inline android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt:207]: I think I figured out what was going on with the [TranslationsMiddlewareTest.kt](https://github.com/mozilla-mobile/firefox-android/pull/5389/files#diff-44da0a222b637b3152adaf0f82484690dd094aad95c4d1760d7266d4f58a6423) pattern over in #5389. That test now seems to be verifying as expected and I think could be used as a pattern for an additional test on this patch too.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt [hunk 4]
```diff
@@ -969,6 +969,17 @@ sealed class TranslationsAction : BrowserAction() {
         override val tabId: String,
         val pageSettings: TranslationPageSettings?,
     ) : TranslationsAction(), ActionWithTab
+
+    /**
+     * Sets the never translate sites.
+     *
+     * @property tabId The ID of the tab the [EngineSession] that requested the list.
+     * @property neverTranslateSites The never translate sites.
```
Comment (ohall-m) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:977]: nit: Maybe a bit more detail like `The list of sites that the user has opted to never translate.` or something.

**Review by ohall-m — APPROVED** [review]
Thanks for working on translations! 

Just a few minor comments about things that recently changed and a suggested pattern for the `MiddlewareTest`.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt [hunk 5]
```diff
@@ -36,4 +36,5 @@ data class TranslationsState(
     val pageSettings: TranslationPageSettings? = null,
     val translationError: TranslationError? = null,
     val settingsError: TranslationError? = null,
+    val neverTranslateSites: List<String>? = null,
```
Comment (ohall-m) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt:39]: nit: Please add the `@property` too, maybe also move above the property before the `Errors`?

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt (hunk 4 above)
Comment (calumozilla) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:977]: Updated the property comment to be more detailed

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt (hunk 1 above)
Comment (calumozilla) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Gotcha, removed `scope.launch`

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt [hunk 6]
```diff
@@ -111,6 +116,44 @@ class TranslationsMiddleware(
...
+                    )
+                    logger.info("Success requesting never translate sites.")
+                },
+
+                onError = {
+                    context.store.dispatch(
+                        TranslationsAction.TranslateExceptionAction(
+                            tabId = tabId,
+                            operation = TranslationOperation.FETCH_LANGUAGES,
+                            translationError = TranslationError.CouldNotLoadLanguagesError(it),
```
[the remark this answers is not in the document]
Comment (calumozilla) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Added `CouldNotLoadNeverTranslateSites`

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt (hunk 5 above)
Comment (calumozilla) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt]: Added documentation and reordered the priorities

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt (hunk 3 above)
Comment (ohall-m) [reply android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt:207]: Awesome, thanks and thanks for the review! https://github.com/mozilla-mobile/firefox-android/pull/5389 just merged!

**Pull request merged**